### PR TITLE
Kernel/IPC: Add a small delay after each SyncRequest to prevent thread starvation.

### LINKED ISSUE #2794: Moco Moco Friends - softlock on chapter screen
Citra version: Nightly - ecd332b

The game had a solflock at the chapter screen. The music was still playing, but inputs doesn't do anything. 

screen shot 06-20-17 at 01 28 pm

[log.txt](https://github.com/citra-emu/citra/files/1090086/log.txt)

System Information
* Operating System: Windows 7 Ultimate Service Pack 1
* CPU: Intel(R) Core(TM) i7-3610QM CPU @ 2.3GHz
* GPU: Geforce GTX 660M
* Citra Version (found in title bar): citra-windows-msvc-20170619-ecd332b
* Game: Moco Moco Friends (USA)

### PR DESCRIPTION
**Pull request by Subv**
In a real 3DS, threads that call svcSyncRequest are put to sleep until the server responds via svcReplyAndReceive. Our HLE services don't implement this mechanism and are effectively immediate from the 3DS's point of view. This commit makes it so that we at least simulate the IPC delay.

Specific HLE handlers might need to put their callers to sleep for a longer period of time to simulate IO timings. This is their responsibility but doing so is currently not implemented.

See https://gist.github.com/ds84182/4a7690c5376e045cab9129ca4185bbeb for a test that was not passing before this commit.

Thanks @ds84182 for pointing this out!

Closes #2794 #1105 #3054 #2647 #2802 #2608 #2749 #2118 #2526 #2749 

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3091)

### CODE REVIEW

**Comment by mailwl** [comment]
looks like fixes Animal Crossing #1105

**Comment by mailwl** [comment]
don't think each SyncRequest needs own delay, because games works well on differ variants of (new/old)3ds/2ds and firmwares

**Comment by mailwl** [comment]
also fixes FIFA (at least 12), but first need we need to signal y2r::completion_event ASYNC

**Comment by Dragios** [comment]
Does it also fix the the missing models on the character / hand in ANCL?

**Comment by mailwl** [comment]
@Dragios, yes. i don't play enough yet, but at start missing chars visible

**Comment by dampih** [comment]
This PR also fixes #3054  .My friend said it also fixes **Dragon Quest XI** stuck after choosing New Game and became playable, but unfortunately it also make **Attack of the Friday Monster! A Tokyo tale** stuck at intro logo. Sorry I don't have those game so I can't test it out.
citra-qt 2017-11-08 16-07-06-14

**Comment by acnleditor2** [comment]
OMG

**Comment by wwylele** [comment]
IMO the default delay time should be the minimal one a `SendSyncRequest` can take. i.e. it is the time taken by a no-op service function. Is this measurable or calculable?

**Comment by acnleditor2** [comment]
this pr causes more problems with mario maker, pigeon dont eats with priority boost hack

**Comment by wwylele** [comment]
@acnleditor2 no other hack talk is allowed here, thanks.

**Comment by Subv** [comment]
@wwylele Agreed. We could write a custom service that has a single function that takes an input ticks value, calculates the difference with the tick value when it is executed, and returns that in the command response. Then we just call that a lot of times and average it out.

**Comment by mailwl** [comment]
just check random softlock game: Moco Moco Friends - softlock on chapter screen #2794
it works.
looks like maaaany issues will be closed

**Comment by dampih** [comment]
Changing initial value `100 `to `90000` give additional bonus fixes. Fixes #2446  #2802  #2745

**Comment by Subv** [comment]
That delay really has to be measured on real hardware, but I believe an IPC overhead of `0.09ms` (`90000ns`) isn't out of the realm of possibility.

**Comment by Dragios** [comment]
Fixes #3061 as well. I have set the delay to 0.1ms.
Pokédex 3D Pro now boots up instantly rather than stuck at black screen.

File: src/core/hle/kernel/server_session.cpp [hunk 1]
```diff
@@ -62,13 +62,28 @@ ResultCode ServerSession::HandleSyncRequest(SharedPtr<Thread> thread) {
...
+        if (hle_handler != nullptr) {
+            // For HLE services, we put the request threads to sleep for a short duration to
+            // simulate IPC overhead, but only if the HLE handler didn't put the thread to sleep for
+            // other reasons like an async callback. The IPC overhead is needed to prevent
+            // starvation when a thread only does sync requests to HLE services while a
+            // lower-priority thread is waiting to run.
+
+            // TODO(Subv): Figure out a good value for this.
+            static constexpr u64 IPCDelayNanoseconds = 100;
+            thread->WakeAfterDelay(IPCDelayNanoseconds);
```
Comment (Subv) [inline src/core/hle/kernel/server_session.cpp]: I just noticed that due to the way WakeAfterDelay is implemented, we can't have delays of less than a microsecond, this is essentially the same as SleepThread(0)

**Comment by Subv** [comment]
I have updated the delay to 54.14 microseconds as measured by @jroweboy. Could you all please re-test everything and report here what was fixed / what was broken by this?

**Comment by bakugo** [comment]
Can confirm that https://github.com/citra-emu/citra/issues/2802 is fixed with this value.

**Comment by acnleditor2** [comment]
Mario Maker: Pigeon stops here. needs a higher value
super mario maker 3ds

**Comment by acnleditor2** [comment]
Pokémon Sun & Moon: freeze after changing the form of Lycanroc in the Pokédex

**Comment by acnleditor2** [comment]
Pokémon Sun & Moon: freeze exiting festival plaza
Pokémon Sun & Moon: freeze after exiting pokedex, entering in festival plaza.

**Comment by Hexagon12** [comment]
Can confirm that #2608 is also fixed.

**Comment by acnleditor2** [comment]
AC:HHD freeze doing a house (except for lottie and goldie)

**Comment by RavenHome1** [comment]
causes pokemon moon to show a black screen instead of the intro

**Comment by acnleditor2** [comment]
achhd freeze on search items by name/color? with some items (only with shader jit enabled)?
achhd freeze on enter in building with shader jit disabled
achhd...................................... is unplayable

**Comment by Chrisatsinnoh** [comment]
Where can I download the fixes, though?

**Comment by bakugo** [comment]
Update on https://github.com/citra-emu/citra/issues/2802: The freeze that always happens when starting the game OR loading a save is indeed fixed with `IPCDelayNanoseconds = 51140`, however it still happens in other situations: if you create a character, let it save, then quit and start again, it will happen after the intro cutscene. Increasing the value further (to `851140`) makes the game not freeze but get stuck on a white screen instead with sound playing.

Seems like this is a pretty complex issue.

**Comment by einstein95** [comment]
This fixes #2749. Star Fox 64 3D now properly initialises save data without needing blank files or premade save files.

**Comment by j-selby** [comment]
Fixes #2118. Woop!

**Comment by AuraOfTheDawn** [comment]
Megaman Legacy collection no longer hangs after the intro. ( #2526 )

**Comment by SigmaVirus** [comment]
I saw a mention of https://github.com/citra-emu/citra/issues/2526, this particular issue has a reference to https://github.com/citra-emu/citra/issues/2238

Might https://github.com/citra-emu/citra/issues/2238 be solved by this PR too by any chance? Any one knows?

**Comment by acnleditor2** [comment]
#3119 breaks ever oasis fix

**Comment by wwylele** [comment]
I am going to temporarily pull this from canary and push #3119 first. We need to make sure it doesn't break anything from master. After everything settle down, we will revisit this fix

**Comment by einstein95** [comment]
@SigmaVirus No, it doesn't

**Comment by bunnei** [comment]
Can this be merged? (I have hesitation since I don't see any more approvals...)

**Comment by Subv** [comment]
This is waiting on someone actually measuring the delay properly

**Comment by mailwl** [comment]
hmm... i suppose New 3ds faster, than 2ds, sorry for trolling

**Comment by B3n30** [comment]
@mailwl nope 2ds is faster.

HW tests with https://gist.github.com/ds84182/ecdbbd25b56a29bd4e5b32a7544b8e92 compared with testing in citra showed that a delay of 39ms will give the same results as the average on new3ds. Those were also values observed on 2ds. Thus a delay of 39ms is a good start. But I also would add a comment on how we obtained that value. Besides changing the delay: LGTM

File: src/core/hle/kernel/server_session.cpp (hunk 1 above)
Comment (tokumeiwokiboushimasu) [reply src/core/hle/kernel/server_session.cpp]: Is this really 51140 instead of 54140?

**Pull request merged**